### ISSUE TITLE
"Invalid sample rate" using JACK w/ PipeWire @ 44.1 kHz

### ISSUE THREAD

**Issue by echozio** [mixxxdj/mixxx#12210]
### Bug Description

I'm getting the following message when starting Mixxx using the JACK backend with PipeWire v0.3.83:

![image](https://github.com/mixxxdj/mixxx/assets/7262891/7f703d05-7257-4409-a88a-00162e15f741)

My hardware as well as other applications use 44.1 kHz, but Mixxx seems determined to use 48 kHz, or at least that's what it sets in its config, even if I change it to 44.1 kHz.

The only way I've managed to make it work is to open a 48 kHz project in Ardour (also using JACK). Then Mixxx doesn't complain, but playback slows down if I close the Ardour session (sample rate changes from 48k to 44.1k according to Carla).

Everything worked until a day or two ago, and only broke either after a system update or after starting to use my DJ controller as the primary output (I believe the sound card I used before always ran at 48 kHz).

### Version

2.3.6

### OS

Debian Sid

**Comment by echozio** [comment]
Some strange things happened after I tested the things mentioned above:
* After closing Ardour and restarting Mixxx, everything seemed to work fine.
* After a reboot Mixxx started just fine
* After starting other applications (slack, spotify, firefox, discord, whatsdesk, session) Mixxx again showed the same error.
* After closing spotify, Mixxx started without issues
* After opening spotify again, Mixxx also started without issues

I don't really know what to make out of this, but will leave it open in case someone deems it worth looking further into. For my own purposes I guess it appears I'm able to work around it by doing random things.

Here's the log from one of the times it didn't start properly: https://gist.github.com/echozio/eef8c6801263eaabe77d87d8e74c2fcd

**Comment by Swiftb0y** [comment]
Can reproduce reliably as well, though I wasn't able to find any time to look into the issue. It's definitely related to a recent pipewire upgrade. My current workaround is just using ALSA again, though that's not great either.

**Comment by Swiftb0y** [comment]
My current hypothesis is that Mixxx always tried to open a device with Jack with the wrong samplerate. Until recently it has worked because pipewire just silently resampled the signal. This is a bug on our part that.

**Comment by Swiftb0y** [comment]
I think the audio slowdown issue is from the fact that mixxx doesn't respond well to pipewires dynamic adjustment of the audio graph. In any case. I'm not sure how to fix this properly. You may be able to work around this by forcing pipewire to stay on 48 KHz no matter what.

**Comment by Swiftb0y** [comment]
requesting help from @daschuer. Why do we / does portaudio not adapt the samplerate to the rate of the jack graph?

**Comment by daschuer** [comment]
You find some hints here: 
https://github.com/PortAudio/portaudio/blob/24c8d575e588d557d28f4011becb753421346860/src/hostapi/jack/pa_jack.c#L616C2-L616C2

In JACK, there is one server with one sample rate.

In case of Pipewire which implements this server interface you need to adjust the sample rate of this "Server" before Mixxx starts: 
`pw-metadata -n settings 0 clock.force-rate 44100`

Portaudio cannot do it, because it dos not know that it is not talking to a real JACK server. 

Portaudio has recently been extended with a Pulse API. This will be able to set the Pipewire samlple rate. There is however some pending work to make that usable for all Mixxx users: 
https://github.com/PortAudio/portaudio/pull/336

**Comment by daschuer** [comment]
Yes, that was the reason we have offered a GSoC project that targets this: 
https://github.com/mixxxdj/mixxx/wiki/GSOC-2023-Ideas#first-class-pipewire-support

The main issue with Portaudio is its device based model, that does not allow full Jack/Pipewire support. So I assume that becoming a native Pipewire (or Jack) app is the solution with the least hassle.

**Comment by Swiftb0y** [comment]
I agree. At that point its probably easier to switch to RtAudio across the board as it follows the channel-oriented model IIUC.

**Comment by eXWoLL** [comment]
Still getting this when trying a mixed ALSA controller/laptop setup.
